File: app/src/main/java/com/aslansari/hypocoin/ui/adapters/CurrencyDiffCallback.java
package com.aslansari.hypocoin.ui.adapters;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DiffUtil;

import com.aslansari.hypocoin.repository.model.Currency;

import java.util.List;

public class CurrencyDiffCallback extends DiffUtil.Callback {

    private final List<Currency> newCurrencyList;
    private final List<Currency> oldCurrencyList;

    public CurrencyDiffCallback(List<Currency> newCurrencyList, List<Currency> oldCurrencyList) {
        this.newCurrencyList = newCurrencyList;
        this.oldCurrencyList = oldCurrencyList;
    }

    @Override
    public int getOldListSize() {
        return oldCurrencyList.size();
    }

    @Override
    public int getNewListSize() {
        return newCurrencyList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return newCurrencyList.get(newItemPosition).id.equals(oldCurrencyList.get(oldItemPosition).id);
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        double newUsdPrice = newCurrencyList.get(newItemPosition).metrics.marketData.priceUSD;
        double oldUsdPrice = oldCurrencyList.get(oldItemPosition).metrics.marketData.priceUSD;
        return newUsdPrice - oldUsdPrice < 0.01;
    }

    @Nullable
    @Override
    public Object getChangePayload(int oldItemPosition, int newItemPosition) {
        return super.getChangePayload(oldItemPosition, newItemPosition);
    }
}

File: app/src/main/java/com/aslansari/hypocoin/ui/adapters/CurrencyDiffCallback.kt
package com.aslansari.hypocoin.ui.adapters

import androidx.recyclerview.widget.DiffUtil
import com.aslansari.hypocoin.repository.model.Currency

class CurrencyDiffCallback(
    private val newCurrencyList: List<Currency>,
    private val oldCurrencyList: List<Currency>
) : DiffUtil.Callback() {
    override fun getOldListSize(): Int {
        return oldCurrencyList.size
    }

    override fun getNewListSize(): Int {
        return newCurrencyList.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return newCurrencyList[newItemPosition].id == oldCurrencyList[oldItemPosition].id
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val newUsdPrice = newCurrencyList[newItemPosition].metrics!!.marketData!!.priceUSD
        val oldUsdPrice = oldCurrencyList[oldItemPosition].metrics!!.marketData!!.priceUSD
        return newUsdPrice - oldUsdPrice < 0.01
    }

    override fun getChangePayload(oldItemPosition: Int, newItemPosition: Int): Any? {
        return super.getChangePayload(oldItemPosition, newItemPosition)
    }
}